src/pybind/mgr/volumes/fs/clone_stats.py

```diff
@@ -0,0 +1,215 @@
+                for ce in clone_index_entries:
+                    ci = CloneInfo(volname)
+
+                    try:
+                        self._get_clone_dst_info(fs_handle, ci, ce,
+                                                 clone_index_path)
+                        self._get_clone_src_info(fs_handle, ci)
+                    except ObjectNotFound as e:
+                        log.info('Exception ObjectNotFound was raised. Apparently '
+                                 'entry in clone index was removed because one of '
+                                 'the clone job(s) has completed/cancelled, '
+                                 'therefore ignoring and proceeding. '
+                                 f'Printing the exception: {e}')
+                        continue
+                    except VolumeException as e:
+                        if e.error_str != 'error fetching subvolume metadata':
+                            raise
+                        log.info('Exception VolumeException was raised. Apparently '
+                                 'an entry from the metadata file of clone source '
+                                 'was removed because one of the clone job(s) has '
+                                 'completed/cancelled. Therefore ignoring and '
+                                 f'proceeding Printing the exception: {e}')
+                        continue
+
+                    if not ci.src_path or not ci.dst_path:
+                        continue
+
+                    clones.append(ci)
+
+        log.debug('finished collecting info on all clones, found '
+                  f'{len(clones)} clones')
+        return clones
+
+    def _update_progress_bars(self):
+        '''
+        Look for amount of progress made by all cloning operations and prints
+        progress bars, in "ceph -s" output, for average progress made
+        accordingly.
+
+        This method is supposed to be run only by instance of class RTimer
+        present in this class.
+        '''
+        clones = self._get_info_for_all_clones()
+        if not clones:
+            self.finish()
+            return
+
+        # onpen bar (that is progress bar for clone jobs in ongoing and pending
+        # state) is printed when clones are in pending state. it is kept in
+        # printing until all clone jobs finish.
+        show_onpen_bar = True if len(clones) > self.max_concurrent_clones \
+            else False
+
+        percent = 0.0
+
+        assert self.on_pev_id is not None
+        sum_percent_ongoing = 0.0
+        avg_percent_ongoing = 0.0
+        total_ongoing_clones = min(len(clones), self.max_concurrent_clones)
+
+        if show_onpen_bar:
+            assert self.onpen_pev_id is not None
+            sum_percent_onpen = 0.0
+            avg_percent_onpen = 0.0
+            total_onpen_clones = len(clones)
+
+        for clone in clones:
+            with open_volume_lockless(self.volclient, clone.volname) as \
```

Comment:
Forgot to post this, this code wasn't added in this PR but moved from `stats_util.py` to `clone_stats.py`. Therefore I don't intend to make improvements on this PR since focus here is add to new feature. Marking this as resolved.